### DISCUSSION TITLE
Error sharing ref usage

### DISCUSSION THREAD

**Discussion by spidertwin2** [react-hook-form/react-hook-form#4891] [Q&A] [answered]
Hello,

I want to create a custom textarea component that grows/shrinks with its content.  But, am having trouble sharing ref usage.  The textarea is flickering with each input change.

[Here](https://codesandbox.io/s/async-wildflower-qnwhb?file=/AutoTextarea.tsx) is a sandbox demo.

Thank you :dolphin:

**Comment by bluebill1049** [comment]
<img width="695" alt="Screen Shot 2021-04-21 at 5 23 06 pm" src="https://user-images.githubusercontent.com/10513364/115513323-492f6800-a2c6-11eb-94f1-de0f4eda6116.png">

try to register before render and then pass the rest prop, and invoke `ref` callback

**Reply by spidertwin2** [reply]
This helps me correctly assign the value for title, but the resize is still abruptly shrinking and growing with each input.

**Reply by bluebill1049** [reply]
would you like to set put this in a separate comment and mark it as an answer, it will help others. 🙏

**Comment by spidertwin2** [comment, marked as answer]
Finally figured it out! :smile: 

Had to use `useLayoutEffect` instead of `useEffect`.  This prevents the textarea from flickering/shrinking on each input change.  Here is the updated code.

AutoTextarea.tsx

```
import { useState, useRef, forwardRef, useLayoutEffect } from "react"
import { useFormContext } from "react-hook-form"
import { Textarea } from "@chakra-ui/react"

const AutoTextarea = forwardRef<HTMLTextAreaElement>(
  (props, reference) => {
    const {
      register,
    } = useFormContext()

    const { ref, ...rest } = register("title")

    const textareaRef = useRef<HTMLTextAreaElement>(null)
    const [currentValue, setCurrentValue] = useState("")

    useLayoutEffect(() => {
      textareaRef.current.style.height = "inherit"
      textareaRef.current.style.height = `${textareaRef.current.scrollHeight}px`
    }, [currentValue])

    const handleChange = (e: any) => {
      e.preventDefault()
      setCurrentValue(e.target.value)
    }

    return (
        <Textarea
          {...rest}
          {...props}
          rows={1}
          name="title"
          ref={(elm: any) => {
            if (elm) {
              ref(elm)
            }
            textareaRef.current = elm
          }}
          onChange={handleChange}
          value={currentValue}
          overflow="hidden"
          resize="none"
          width="100%"
        />
    )
  }
)

export default AutoTextarea
```

**Reply by bluebill1049** [reply]
thanks!
